core/src/main/scala/kafka/controller/PartitionStateMachineV2.scala

```diff
@@ -0,0 +1,268 @@
+      val liveReplicas = controllerContext.partitionReplicaAssignment(partition)
+        .filter(replica => controllerContext.isReplicaOnline(replica, partition, true))
+      val isr = leaderIsrAndControllerEpoch.leaderAndIsr.isr
+      val leaderOpt = PartitionLeaderElectionAlgorithms.controlledShutdownPartitionLeaderElection(isr, liveReplicas.toSet, shuttingDownBrokers)
+      val newIsr = isr.filter(replica => !controllerContext.shuttingDownBrokerIds.contains(replica))
+      val newLeaderAndIsrOpt = leaderOpt.map(leader => leaderIsrAndControllerEpoch.leaderAndIsr.newLeaderAndIsr(leader, newIsr))
+      val nextLeaderIsrAndControllerEpochOpt = newLeaderAndIsrOpt.map(newLeaderAndIsr => LeaderIsrAndControllerEpoch(newLeaderAndIsr, controller.epoch))
+      (partition, nextLeaderIsrAndControllerEpochOpt, liveReplicas)
+    }
+  }
+
+  private def isValidTransition(partition: TopicAndPartition, targetState: PartitionState) =
+    targetState.validPreviousStates.contains(partitionState(partition))
+
+  private def logInvalidTransition(partition: TopicAndPartition, targetState: PartitionState): Unit = {
+    val currState = partitionState(partition)
+    val e = new IllegalStateException("Partition %s should be in the %s states before moving to %s state"
+      .format(partition, targetState.validPreviousStates.mkString(","), targetState) + ". Instead it is in %s state"
+      .format(currState))
+    logFailedStateChange(partition, currState, targetState, e)
+  }
+
+  private def logFailedStateChange(partition: TopicAndPartition, currState: PartitionState, targetState: PartitionState, t: Throwable): Unit = {
+    stateChangeLogger.error("Controller %d epoch %d initiated state change for partition %s from %s to %s failed"
+      .format(controllerId, controller.epoch, partition, currState, targetState), t)
+  }
+}
+
+object PartitionLeaderElectionAlgorithms {
+  def offlinePartitionLeaderElection(assignment: Seq[Int], isr: Seq[Int], liveReplicas: Set[Int], uncleanLeaderElectionEnabled: Boolean): Option[Int] = {
+    assignment.find(id => liveReplicas.contains(id) && isr.contains(id)).orElse {
+      if (uncleanLeaderElectionEnabled) {
+        assignment.find(liveReplicas.contains)
+      } else {
+        None
+      }
+    }
+  }
+
+  def reassignPartitionLeaderElection(reassignment: Seq[Int], isr: Seq[Int], liveReplicas: Set[Int]): Option[Int] = {
+    reassignment.find(id => liveReplicas.contains(id) && isr.contains(id))
+  }
+
+  def preferredReplicaPartitionLeaderElection(assignment: Seq[Int], isr: Seq[Int], liveReplicas: Set[Int]): Option[Int] = {
+    assignment.headOption.filter(id => liveReplicas.contains(id) && isr.contains(id))
+  }
+
+  def controlledShutdownPartitionLeaderElection(isr: Seq[Int], liveReplicas: Set[Int], shuttingDownBrokers: Set[Int]): Option[Int] = {
+    isr.find(id => liveReplicas.contains(id) && !shuttingDownBrokers.contains(id))
```

Comment:
Ideally, we want to select the preferred replica if it's alive, in-sync and not shutting down as ControlledShutdownLeaderSelector does.